### DISCUSSION TITLE
添加辅助码之后的候选字排序似乎有 bug？

### DISCUSSION THREAD

**Discussion by BettyJJ** [rime/home#751] [Q&A]
我构造了一个最简例子来展示这个疑似 bug，schema 和码表见 https://gist.github.com/BettyJJ/7a837441b80f485082102ef5ea2af7d7
示例码表中只包含读 tai 的字，给每个字拼音后面加上了辅助码，比如“太”成为 `tai;dd`

输入 tai，翻一两页，会看到很多生僻字排序很靠前

如果把码表中从分号开始的辅助码部分全都去掉，只保留全拼
重新部署后再输入 tai，翻一两页，可以看到前面都是常见字

因为是最简例子，所以没加任何拼写运算（也试过加各种拼写运算，对这个效果没影响）
码表中没有词频，是 `use_preset_vocabulary:	true` 导入了预设的 essay.txt，而 essay.txt 文件我没改过，点开看了，里面的词频都是正常的

我还试验过把所有字的辅助码都写成一样，比如都写成 `tai;pp`，这时输入 tai 就发现排序正常，跟只保留全拼效果一样

所以，看起来似乎是辅助码本身的码影响了排序
我还试过把一个辅助码是 dd 的字改成 zd，它在候选字里的顺序就从很前面跑到后面去了……

这是个 bug？还是什么隐藏机制？还是我哪里没配置对？
应如何解决？

**Comment by lotem** [comment]
​示例的方案代碼沒有定義拼寫運算規則。以編碼術的話語講，是沒有定義什麼是「主碼」。

因此沒有輸入完整編碼的時候，所有的候選字都是根據未完成的那部分編碼預測出來的，這類結果排序按照編碼的字典順序排列。

**Reply by BettyJJ** [reply]
感谢回复~
也试过加各种拼写运算，对这个效果没影响啊~
想要达到的效果是打不打辅助码都能打出字，比如如下这么写，是都能出字但排序是乱的

     - derive/^(\w+);\w{2}$/$1/         #不打辅助码也可以匹配
     - derive/^(\w+);(\w)\w$/$1;$2/     #只打一位辅助码也可以匹配

> 這類結果排序按照編碼的字典順序排列

既然指定了 `sort:	by_weight` 就应该按词频排序，不应该默默地又变成字典排序吧……
而且事实上 Rime 好像也不是这么做的……对了，“编码的字典顺序”是指什么？是 dict.yamal 文件里的条目顺序么？还是编码的 abcd 顺序？不管哪种，跟这个示例的结果都不符合……
而且如果编码都写成 tai;pp，这时输入 tai 也是不完整编码，但排序就是正常的啊……

总而言之，要如何才能指示 Rime 在这种情况下按词频排序呢？

**Reply by lotem** [reply]
前文「編碼的字典順序」是指字母表順序。
`sort: by_weight` 是指重碼字之間的排序方法。（重碼：通過詞典及拼寫運算所定義的編碼相同）
對未完整輸入的編碼不能按詞頻排序，因爲這種情況（比如只輸入編碼的第一個字母時）程序只檢索出足夠顯示在當前頁的候選字，無法在所有編碼前綴匹配的候選之間比較詞頻。

**Reply by BettyJJ** [reply]
那，我想要打不打辅助码都能打出字，并且排序按照词频，就没有办法能做到了吗？😿
我是按佛振大佬这个帖子做的：https://tieba.baidu.com/p/2094178562 大佬没提过排序会乱掉这回事呀

**Reply by lotem** [reply]
沒有這回事。是你沒做出來。我的帖子沒有說只要輸入任意一部分編碼就能按字頻順序枚舉。

完整輸入字典定義的、或通過拼寫運算推導出來的編碼，候選字會默認按照字頻排序。

實例：[地球拼音](https://github.com/rime/rime-terra-pinyin)輸入不帶聲調的拼音（通過拼寫運算由詞典中包含聲調的拼音推導出來），類比只輸入主碼。聲調可類比輔助碼。

**Reply by lotem** [reply]
> 想要打不打辅助码……

你得告訴輸入法什麼是輔助碼。並不是隨便省略一部分編碼都行。

**Reply by BettyJJ** [reply]
那能不能说一下应该怎么写啊？
前面说了我试了写这样的运算规则 `- derive/^(\w+);\w{2}$/$1/` 
我理解就是 tai;pp 推导出 tai 呀（; 后面是辅助码，辅助码一定是两位，前面的拼音不定长），可是这样输入 tai 排序还是乱的
请教一下到底应该怎么写呢？感谢

**Reply by lotem** [reply]
​做法我在 https://tieba.baidu.com/p/2094178562 演示得很清楚，代碼也詳細解說了。

我不知道你具體應該怎麼寫，我對你的方案不熟。
你總不能都讓我幫你完成吧？我只發佈輸入法框架軟件，不包攬定製服務。

**Reply by BettyJJ** [reply]
呃，就是我这个示例方案，能否具体指出我上面这个运算规则有什么错呢？
大佬能回复已经很感谢了，不敢求什么定制服务
但是按照我的理解，我已经“完整輸入了通過拼寫運算推導出來的編碼”，却没有按照词频排序，会不会还是有什么隐藏 bug 呢？
大佬一句话，抵得上我们摸索几十几百个小时，哭泣恳请赐教 T_T

**Reply by lotem** [reply]
​效果不對肯定有BUG啊，比方說拼寫運算沒生效。
你還是具體調試一下吧，加個前綴後綴`s/( )/x$1y/` 之類驗證一下結果。我也沒看出有什麼問題。

**Comment by BettyJJ** [comment]
@lotem  报告，又经过多次反复试验，发现了一个更奇怪的现象……
我找了别人一份能正常用的方案，提取出读 tai 的字进行试验，还是在我这个 gist：https://gist.github.com/BettyJJ/7a837441b80f485082102ef5ea2af7d7
按照这个 gist 现有的内容部署，输入 tai，候选字排序正常
![image](https://user-images.githubusercontent.com/245282/111716013-18f04600-8890-11eb-903a-60f03607c27f.png)

但是，只要把“台”字的辅助码改一下，本来是 sz，改一个字母比如改成 sx，排序就变了！会有一个奇怪的生僻字出现在第一页
![image](https://user-images.githubusercontent.com/245282/111716037-273e6200-8890-11eb-8b70-8e1c42f4492f.png)
方案的其他任何地方都没有变！只是把“台”这个字的编码改一个字母，就会影响排序，而且甚至都不是影响“台”自己的排序，而是出现了一个毫无关系的生僻字
改成 sa、st 之类也会出现这个生僻字，但改成 ss 就是正常的
这已经超出了我的理解范围，进入玄学领域了……简直感觉我是不是得罪了输入法之神 😭
难道这也是拼写运算规则不对吗……附上现有规则，从别人能用的方案里抄过来的：
```
speller:
  alphabet: "abcdefghijklmnopqrstuvwxyz"
  initials: "abcdefghijklmnopqrstuvwxyz"
  delimiter: " '"
  algebra:
    - derive/^(..).*$/$1/
    - derive/^(...).*$/$1/
    - derive/^(....).*$/$1/
    - derive/^(.....).*$/$1/
```

占用大佬很多时间，很抱歉，感谢您一直回复。
折腾 Rime 配置我也花了好多时间和心血，现在遇到这种事，觉得好难过……可能我和它没有缘分吧，唉 T_T

**Comment by LEOYoon-Tsaw** [comment]
試試
derive/^([a-z]+);[a-z]+$/$1/
derive/^([a-z]+);([a-z])[a-z]*$/$1$2/
xform/^([a-z]+);([a-z]+)$/$1$2/

**Comment by functoreality** [comment]
我遇到过的一个问题与这个讨论的标题描述有点类似，虽然和事实上讨论的东西不一样。鉴于可能有用户遇到与我同样的问题而点进这个帖子，我还是在这里简单描述一下，以便参考。不作为本问题的回答。（如果我的回复更适合放在其他某个地方，可以告诉我。）
例如，某个生僻字编码 `aa`，某个常见字编码 `aa;bb`，则输入 `aa` 时，我的经验是 Rime 会优先展示完全匹配的字（也就是那个生僻字），再展示需要自动补全、或者通过拼写运算匹配的字（那个常用字），这个优先级高于词频。感觉这样设计也还算合理，因为补全的方式有很多，靠词频优先排序可能会导致那个生僻字需要翻页很多次才能找到，而因为已经是完全匹配，继续输入也无法让它排到前面；而常用字只需要多打几个字符就能排到前面。
对此我的解决方法是，把生僻字的编码改成 `aa;;`，避免输入 `aa` 时发生完全匹配。相应的拼写运算并不难写。
